### DISCUSSION TITLE
Curious why is_not_null doesn't work in this query?

### DISCUSSION THREAD

**Discussion by u9g** [obi1kenobi/trustfall#277] [Q&A] [answered]
Was messing around on the query playground, but this query seems to fail:
```
query {
  Latest {
    comment {
      reply @recurse(depth: 10) {
        byUser {
          link {
            url @filter(op: "is_not_null") @output
          }
        }
      }
    }
  }
}
```
with error:
```Error:
Incompatible types encountered in @filter.
```

I also found this query that uses `is_not_null` that fails with the same error: https://github.com/obi1kenobi/trustfall/blob/main/trustfall/examples/hackernews/README.md#example-jobs-in-the-top-50-items . Maybe a problem with `is_not_null`, not sure. Is there some other way to do this that I am not seeing?

Anyways, I'm having a blast messing around with queries, thanks for releasing such a cool tool, would love to hear back.

**Comment by obi1kenobi** [comment, marked as answer]
I'm sorry, that's a terrible error message. It's supposed to show the underlying error text as well — fixing in #278.

In the playground query you posted, that `url` property is defined as non-nullable in the schema — note the `!` in its type:
<img src="https://github.com/obi1kenobi/trustfall/assets/2348618/e1af0cd0-ee71-4ce4-b4c6-72adeb9e2126" height="200">

Filtering for non-null doesn't really make sense on properties that are already guaranteed to be non-null by the schema, hence the error.

Thanks for flagging the bad example query as well. If you'd like to make a PR to fix it, I'd happily merge it and you'll get a shout-out in the next release notes for the contribution :)

**Reply by u9g** [reply]
Oh, that makes more sense, although, I think that is indeed nullable, judging by the fact that I just got a null: 
![image](https://github.com/obi1kenobi/trustfall/assets/43508353/5f2f4ba9-3654-4d03-9e00-6d5b1e421d73)

**Reply by obi1kenobi** [reply]
That's ... exciting 😅 Must be a bug in [the adapter implementation for the `link` edge](https://github.com/obi1kenobi/trustfall/blob/main/experiments/browser_based_querying/src/hackernews/adapter.ts#L107-L143).

Could you figure out which username that is for, and reproduce the problem with a query like:
```graphql
{
    User(name: "<name here>") {
        link {
            url @output
        }
    }
}
```

Then it should be easy to figure out what the bug in the adapter is. If you're open to opening a quick PR with the fix, I would gratefully merge it promptly 🙏

**Reply by u9g** [reply]
```
query {
    User(name: "cs702") {
        link {
            url @output
        }
    }
}
```

I'd love to help with a pr but I seem to be having a problem even building the browser_based_querying locally after following the steps in the readme where I just get a blank screen and in the console I get `TypeError: can't access property "__wbindgen_add_to_stack_pointer", wasm is undefined`. Not sure if this is a known problem, but don't see any issues for it at the moment.

**Reply by obi1kenobi** [reply]
Oof. I'll sort out the build problem in the morning (it's past midnight over here right now) and I'll give you a ping here when I've resolved it so you can give it another shot.

Sorry everything is broken today 😞

**Reply by u9g** [reply]
Hey, any updates on this?

**Reply by obi1kenobi** [reply]
No luck yet, still looking. Hoping to resolve by tonight.

**Reply by obi1kenobi** [reply]
I'm hoping it was just a version mismatch where npm was using an older wasm-pack than Rust was, leading to a calling convention mismatch. This is because I wasn't able to reproduce the problem locally, and after upgrading to latest stable npm and latest wasm-pack (both in npm and globally, for Rust), I again can't reproduce the problem.

PR that is hopefully the fix: https://github.com/obi1kenobi/trustfall/pull/279

Can you let me know if that fixed it for you? You'll need to run `npm install` in the `browser_based_querying` directory to pick up the new deps. I also recommend updating to latest LTS node, just to be safe.

**Reply by u9g** [reply]
Yeah, that fixes it, thanks for that! After installing, I was able to find a fix to this bug! https://github.com/obi1kenobi/trustfall/pull/280

**Reply by obi1kenobi** [reply]
In retrospect I should have just opened that PR way earlier instead of continuing to try and try to replicate the problem 😅

Thanks for the fix! I've queued it up and it should get merged shortly.

Out of curiosity, mind if I ask how you found Trustfall and what drew you to it?

**Reply by u9g** [reply]
Yeah absolutely, I saw your tweet about you going to new york and looking for people and then linked this project. I took a look at the project and saw your intro video and was stunned. This is like the coolest thing ever, this is what I think the internet should be about, connecting little bits of data together to form a big web of data that people can use.

In any event, more immediately, I kind of want to make some adapters for some things I mess around with manually in js but having trouble finding my way around the codebase/understanding the edges stuff so I thought I'd mess around more with the playground.

Is there a discord or community somewhere?

**Reply by obi1kenobi** [reply]
Thank you for the kind words 😊 What kinds of stuff are you looking to automate, any examples you don't mind sharing?

I'm looking into setting up a Discord, I'm just worried about spam and such (e.g. what happened to the Rust Community Discord) and it'll be another month or so before I can really dig into the moderation tools enough to feel comfortable making a community server.

In the meantime, I'm in several Rust / systems-themed Discords, so perhaps you can find me there? I'm also on Twitter, Mastodon, and Bluesky, and obviously here in the Discussions tab.

I look forward to seeing what you build! Please post them as show-and-tell items here in Discussions :) And if anything isn't clear in the docs, or in the code, or just in general with the project, please feel free to ask questions and I'll do my best to answer.

**Reply by u9g** [reply]
I'm in college now, so I have to look for new classes each semester. I'd love to be able to crosscheck classes I need with ratemyprofessor grade for the professor in a composable way. Although, to be honest I am just having fun in general adding sources at the moment to the example, as this project is blowing up what I thought was possible with coding.

That's completely understandable, I've added you on discord so I may ask some things over there in the future, and am super excited to build stuff, thanks for making this open source and being so helpful!

**Reply by obi1kenobi** [reply]
Awesome! That's exactly the sort of use case Trustfall is meant for: cases where you "technically have the data" and yet without Trustfall, the next best option would be either a completely bespoke one-off script or just doing it all manually by clicking around a bunch.

I'm happy to help, and I'd be thrilled to see what you build — please do share screenshots, experiences, queries, anything you deem interesting to point out!
